File: server/src/services/terrainAnalysis.js
/**
 * Terrain Analysis Service
 * 
 * Creates an interpolated DEM (Digital Elevation Model) from contour lines
 * using TIN (Triangulated Irregular Network) interpolation.
 * 
 * Computes:
 * - Elevation grid from contour data
 * - Slope and aspect at each cell
 * - D8 flow direction
 * - Flow accumulation
 * 
 * Limitations:
 * - Interpolated DEM is less accurate than surveyed DEM
 * - Flat areas between contour lines may have artifacts
 * - Grid resolution trades off between accuracy and performance
 */

const turf = require('@turf/turf');
const {
  calculateBounds,
  gridToCoords,
  coordsToGrid,
  sampleLinePoints,
  barycentricInterpolation,
  haversineDistance
} = require('../utils/geometry');

// D8 flow direction encoding: 8 neighbors
// Index: 0=E, 1=NE, 2=N, 3=NW, 4=W, 5=SW, 6=S, 7=SE
const DR = [0, 1, 1, 1, 0, -1, -1, -1]; // row offsets
const DC = [1, 1, 0, -1, -1, -1, 0, 1]; // col offsets

/**
 * Build a terrain model from contour features.
 * @param {Array} features - GeoJSON contour line features with elevation
 * @param {Object} metadata - Contour metadata (bounds, elevations)
 * @param {number} targetCellCount - Approximate number of grid cells per axis (default 80)
 * @returns {Object} Terrain model with elevation grid, flow direction, flow accumulation
 */
function buildTerrainModel(features, metadata, targetCellCount = 80) {
  const { bounds } = metadata;

  // Calculate grid dimensions
  const lngRange = bounds.maxLng - bounds.minLng;
  const latRange = bounds.maxLat - bounds.minLat;

  // Determine cell size (in degrees)
  const maxRange = Math.max(lngRange, latRange);
  const cellSize = maxRange / targetCellCount;
  const cellSizeLng = cellSize;
  const cellSizeLat = cellSize;

  const nCols = Math.max(5, Math.ceil(lngRange / cellSizeLng));
  const nRows = Math.max(5, Math.ceil(latRange / cellSizeLat));

  console.log(`  Grid dimensions: ${nRows} rows × ${nCols} cols (cell ~${cellSize.toFixed(6)}°)`);

  // Calculate approximate cell size in meters for reference
  const centerLat = (bounds.minLat + bounds.maxLat) / 2;
  const cellSizeMeters = haversineDistance(
    centerLat, bounds.minLng,
    centerLat, bounds.minLng + cellSizeLng
  );
  console.log(`  Approximate cell size: ${cellSizeMeters.toFixed(1)}m`);

  // Step 1: Sample points from contour lines
  const samplePoints = sampleContourPoints(features, nCols);

  console.log(`  Sampled ${samplePoints.length} points from ${features.length} contour lines`);

  // Step 2: Build TIN from sample points
  const tin = buildTIN(samplePoints);

  // Step 3: Interpolate elevation grid
  const elevationGrid = interpolateGrid(
    tin, samplePoints, nRows, nCols,
    bounds, cellSizeLng, cellSizeLat
  );

  // Step 4: Fill NoData cells using nearest-neighbor
  fillNoData(elevationGrid, nRows, nCols);

  // Step 5: Calculate flow direction (D8)
  const flowDirection = calculateFlowDirection(elevationGrid, nRows, nCols);

  // Step 6: Calculate flow accumulation
  const flowAccumulation = calculateFlowAccumulation(
    elevationGrid, flowDirection, nRows, nCols
  );

  return {
    elevationGrid,
    flowDirection,
    flowAccumulation,
    nRows,
    nCols,
    bounds,
    cellSizeLng,
    cellSizeLat,
    cellSizeMeters
  };
}

/**
 * Sample elevation points from contour lines.
 */
function sampleContourPoints(features, gridSize) {
  const points = [];
  // Limit points per line based on grid size
  const maxPointsPerLine = Math.max(5, Math.floor(gridSize * 0.5));

  for (const feature of features) {
    if (feature.properties.elevation === null) continue;
    const sampled = sampleLinePoints(
      feature.geometry.coordinates,
      feature.properties.elevation,
      maxPointsPerLine
    );
    points.push(...sampled);
  }

  return points;
}

/**
 * Build a TIN (Triangulated Irregular Network) from sample points.
 * Uses Turf.js tin function.
 */
function buildTIN(samplePoints) {
  // Create a Turf point collection
  const pointFeatures = samplePoints.map(p =>
    turf.point([p.lng, p.lat], { elevation: p.elevation })
  );
  const pointCollection = turf.featureCollection(pointFeatures);

  // Generate TIN
  const tin = turf.tin(pointCollection, 'elevation');
  return tin;
}

/**
 * Interpolate elevation at grid points using TIN.
 */
function interpolateGrid(tin, samplePoints, nRows, nCols, bounds, cellSizeLng, cellSizeLat) {
  const grid = Array.from({ length: nRows }, () =>
    new Float64Array(nCols).fill(NaN)
  );

  // Pre-process TIN triangles for faster lookup
  const triangles = tin.features.map(f => {
    const coords = f.geometry.coordinates[0]; // polygon ring
    return {
      v0: { x: coords[0][0], y: coords[0][1], z: f.properties.a },
      v1: { x: coords[1][0], y: coords[1][1], z: f.properties.b },
      v2: { x: coords[2][0], y: coords[2][1], z: f.properties.c },
      bbox: {
        minX: Math.min(coords[0][0], coords[1][0], coords[2][0]),
        maxX: Math.max(coords[0][0], coords[1][0], coords[2][0]),
        minY: Math.min(coords[0][1], coords[1][1], coords[2][1]),
        maxY: Math.max(coords[0][1], coords[1][1], coords[2][1])
      }
    };
  });

  // For each grid cell, find the containing triangle and interpolate
  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      const [lng, lat] = gridToCoords(row, col, bounds, cellSizeLng, cellSizeLat);

      // Search through triangles
      for (const tri of triangles) {
        // Quick bbox check
        if (lng < tri.bbox.minX || lng > tri.bbox.maxX ||
            lat < tri.bbox.minY || lat > tri.bbox.maxY) continue;

        const z = barycentricInterpolation(lng, lat, tri.v0, tri.v1, tri.v2);
        if (z !== null) {
          grid[row][col] = z;
          break;
        }
      }
    }
  }

  return grid;
}

/**
 * Fill NoData (NaN) cells using nearest-neighbor interpolation.
 * Iteratively expands from known cells.
 */
function fillNoData(grid, nRows, nCols) {
  let changed = true;
  let iterations = 0;
  const maxIterations = Math.max(nRows, nCols);

  while (changed && iterations < maxIterations) {
    changed = false;
    iterations++;

    for (let row = 0; row < nRows; row++) {
      for (let col = 0; col < nCols; col++) {
        if (!isNaN(grid[row][col])) continue;

        // Check 8 neighbors
        let sum = 0, count = 0;
        for (let d = 0; d < 8; d++) {
          const nr = row + DR[d];
          const nc = col + DC[d];
          if (nr >= 0 && nr < nRows && nc >= 0 && nc < nCols && !isNaN(grid[nr][nc])) {
            sum += grid[nr][nc];
            count++;
          }
        }

        if (count > 0) {
          grid[row][col] = sum / count;
          changed = true;
        }
      }
    }
  }

  // Any remaining NaN cells, fill with global mean
  let globalSum = 0, globalCount = 0;
  for (let r = 0; r < nRows; r++) {
    for (let c = 0; c < nCols; c++) {
      if (!isNaN(grid[r][c])) {
        globalSum += grid[r][c];
        globalCount++;
      }
    }
  }
  const globalMean = globalCount > 0 ? globalSum / globalCount : 0;
  for (let r = 0; r < nRows; r++) {
    for (let c = 0; c < nCols; c++) {
      if (isNaN(grid[r][c])) grid[r][c] = globalMean;
    }
  }
}

/**
 * Calculate D8 flow direction for each cell.
 * Each cell flows to its steepest downhill neighbor.
 * Returns a 2D array of direction indices (0-7) or -1 for flat/pit cells.
 */
function calculateFlowDirection(elevationGrid, nRows, nCols) {
  const flowDir = Array.from({ length: nRows }, () =>
    new Int8Array(nCols).fill(-1)
  );

  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      const elev = elevationGrid[row][col];
      let maxDrop = 0;
      let bestDir = -1;

      for (let d = 0; d < 8; d++) {
        const nr = row + DR[d];
        const nc = col + DC[d];

        if (nr < 0 || nr >= nRows || nc < 0 || nc >= nCols) {
          // Boundary: allow flow out
          const drop = elev - elevationGrid[row][col]; // flat, but edge
          if (bestDir === -1 && (row === 0 || row === nRows - 1 || col === 0 || col === nCols - 1)) {
            bestDir = d; // Flow to boundary
          }
          continue;
        }

        const neighborElev = elevationGrid[nr][nc];
        // Diagonal distance factor
        const dist = (d % 2 === 0) ? 1.0 : 1.414;
        const drop = (elev - neighborElev) / dist;

        if (drop > maxDrop) {
          maxDrop = drop;
          bestDir = d;
        }
      }

      flowDir[row][col] = bestDir;
    }
  }

  return flowDir;
}

/**
 * Calculate flow accumulation using topological sort.
 * Sorts cells from highest to lowest elevation and accumulates flow downstream.
 */
function calculateFlowAccumulation(elevationGrid, flowDirection, nRows, nCols) {
  const accumulation = Array.from({ length: nRows }, () =>
    new Float64Array(nCols).fill(1) // Each cell starts with 1 (itself)
  );

  // Sort cells by elevation (highest first)
  const cells = [];
  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      cells.push({ row, col, elev: elevationGrid[row][col] });
    }
  }
  cells.sort((a, b) => b.elev - a.elev);

  // Process from highest to lowest
  for (const { row, col } of cells) {
    const dir = flowDirection[row][col];
    if (dir < 0) continue;

    const nr = row + DR[dir];
    const nc = col + DC[dir];

    if (nr >= 0 && nr < nRows && nc >= 0 && nc < nCols) {
      accumulation[nr][nc] += accumulation[row][col];
    }
  }

  return accumulation;
}

module.exports = {
  buildTerrainModel,
  DR,
  DC
};

File: python_server/services/terrain_analysis.py
"""
Terrain Analysis Service
Builds a DEM (Digital Elevation Model) from contour line features.

Steps:
  1. Sample points from contour lines
  2. Interpolate elevation grid using scipy LinearNDInterpolator (fast, vectorised)
  3. Fill NoData cells with nearest-neighbour
  4. Compute D8 flow direction
  5. Compute flow accumulation (topological sort)

D8 direction encoding (same as Node.js version):
  Index: 0=E, 1=NE, 2=N, 3=NW, 4=W, 5=SW, 6=S, 7=SE
"""

import math
import numpy as np
from scipy.interpolate import LinearNDInterpolator, NearestNDInterpolator
from typing import Any

# D8 neighbour offsets — row, col
DR = np.array([0, -1, -1, -1,  0,  1,  1,  1], dtype=np.int8)
DC = np.array([1,  1,  0, -1, -1, -1,  0,  1], dtype=np.int8)

EARTH_RADIUS_M = 6_371_000.0


def build_terrain_model(features: list, metadata: dict, target_cell_count: int = 50) -> dict[str, Any]:
    """
    Build terrain model from contour features.

    Args:
        features: list of GeoJSON-like contour features with elevation
        metadata: contour metadata (bounds, elevations)
        target_cell_count: approximate grid size per axis (higher = more detail, slower)
    Returns:
        dict with elevation_grid, flow_direction, flow_accumulation and grid geometry
    """
    bounds = metadata['bounds']
    lng_range = bounds['maxLng'] - bounds['minLng']
    lat_range = bounds['maxLat'] - bounds['minLat']

    max_range = max(lng_range, lat_range)
    cell_size = max_range / target_cell_count

    n_cols = max(5, math.ceil(lng_range / cell_size))
    n_rows = max(5, math.ceil(lat_range / cell_size))

    cell_size_lng = lng_range / n_cols
    cell_size_lat = lat_range / n_rows

    center_lat = (bounds['minLat'] + bounds['maxLat']) / 2
    cell_size_meters = _haversine(center_lat, bounds['minLng'], center_lat, bounds['minLng'] + cell_size_lng)

    print(f"  Grid: {n_rows}×{n_cols}, cell ~{cell_size_meters:.1f}m")

    # ── Step 1: Sample points ──────────────────────────────────────────────
    points, values = _sample_contour_points(features, n_cols)
    print(f"  Sampled {len(points)} points from {len(features)} contour lines")

    # ── Step 2: Interpolate grid (vectorised scipy) ────────────────────────
    elevation_grid = _interpolate_grid(points, values, n_rows, n_cols, bounds, cell_size_lng, cell_size_lat)

    # ── Step 3: D8 flow direction ──────────────────────────────────────────
    flow_direction = _calc_flow_direction(elevation_grid, n_rows, n_cols)

    # ── Step 4: Flow accumulation ──────────────────────────────────────────
    flow_accumulation = _calc_flow_accumulation(elevation_grid, flow_direction, n_rows, n_cols)

    return {
        'elevation_grid':   elevation_grid,
        'flow_direction':   flow_direction,
        'flow_accumulation': flow_accumulation,
        'nRows':            n_rows,
        'nCols':            n_cols,
        'bounds':           bounds,
        'cellSizeLng':      cell_size_lng,
        'cellSizeLat':      cell_size_lat,
        'cellSizeMeters':   cell_size_meters,
    }


# ─── Grid helpers ────────────────────────────────────────────────────────────

def grid_to_coords(row: int, col: int, bounds: dict, cell_size_lng: float, cell_size_lat: float):
    """Convert grid (row, col) → (lng, lat) centre coordinates."""
    lng = bounds['minLng'] + (col + 0.5) * cell_size_lng
    lat = bounds['minLat'] + (row + 0.5) * cell_size_lat
    return lng, lat


def coords_to_grid(lng: float, lat: float, bounds: dict, cell_size_lng: float, cell_size_lat: float):
    """Convert (lng, lat) → nearest grid (row, col)."""
    col = int((lng - bounds['minLng']) / cell_size_lng)
    row = int((lat - bounds['minLat']) / cell_size_lat)
    return row, col


def _haversine(lat1, lng1, lat2, lng2) -> float:
    """Haversine distance in metres."""
    r = EARTH_RADIUS_M
    dlat = math.radians(lat2 - lat1)
    dlng = math.radians(lng2 - lng1)
    a = math.sin(dlat / 2) ** 2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlng / 2) ** 2
    return 2 * r * math.asin(math.sqrt(a))


# ─── Sampling ────────────────────────────────────────────────────────────────

def _sample_contour_points(features: list, grid_size: int):
    """Sample (lng, lat) points and their elevations from contour lines."""
    max_per_line = max(5, grid_size // 2)
    pts, vals = [], []

    for feature in features:
        elev = feature['properties']['elevation']
        coords = feature['geometry']['coordinates']
        n = len(coords)
        if n == 0:
            continue

        if n <= max_per_line:
            sampled = coords
        else:
            # Evenly spaced indices
            indices = [round(i * (n - 1) / (max_per_line - 1)) for i in range(max_per_line)]
            sampled = [coords[i] for i in indices]

        for lng, lat in sampled:
            pts.append([lng, lat])
            vals.append(elev)

    return np.array(pts, dtype=np.float64), np.array(vals, dtype=np.float64)


# ─── Interpolation ───────────────────────────────────────────────────────────

def _interpolate_grid(points, values, n_rows, n_cols, bounds, cell_size_lng, cell_size_lat) -> np.ndarray:
    """
    Interpolate elevation for every grid cell using scipy LinearNDInterpolator.
    This is O(n log n) and vectorised — orders of magnitude faster than the
    per-triangle loop used in the Node.js implementation.
    """
    # Build grid of all cell-centre coordinates
    cols_arr = np.arange(n_cols)
    rows_arr = np.arange(n_rows)
    lngs = bounds['minLng'] + (cols_arr + 0.5) * cell_size_lng
    lats = bounds['minLat'] + (rows_arr + 0.5) * cell_size_lat
    grid_lngs, grid_lats = np.meshgrid(lngs, lats)  # shape: (n_rows, n_cols)
    grid_pts = np.column_stack([grid_lngs.ravel(), grid_lats.ravel()])

    # Linear (TIN-equivalent) interpolation
    lin_interp = LinearNDInterpolator(points, values)
    elev_flat = lin_interp(grid_pts)

    # Fill any NaN (outside convex hull) with nearest-neighbour
    nan_mask = np.isnan(elev_flat)
    if nan_mask.any():
        nn_interp = NearestNDInterpolator(points, values)
        elev_flat[nan_mask] = nn_interp(grid_pts[nan_mask])

    return elev_flat.reshape(n_rows, n_cols)


# ─── D8 Flow direction ───────────────────────────────────────────────────────

def _calc_flow_direction(elev: np.ndarray, n_rows: int, n_cols: int) -> np.ndarray:
    """
    D8 flow direction: each cell flows to its steepest downhill neighbour.
    Returns int8 array, -1 = no downhill neighbour (pit/boundary).
    """
    flow_dir = np.full((n_rows, n_cols), -1, dtype=np.int8)

    for d in range(8):
        dr, dc = int(DR[d]), int(DC[d])
        dist = 1.414 if d % 2 == 1 else 1.0

        # Shift elevation grid in direction d
        src = elev
        shifted = np.roll(np.roll(elev, -dr, axis=0), -dc, axis=1)

        # Boundary cells that rolled to the other side are invalid
        row_mask = np.ones(n_rows, dtype=bool)
        col_mask = np.ones(n_cols, dtype=bool)
        if dr > 0:
            row_mask[-dr:] = False
        elif dr < 0:
            row_mask[:-dr] = False
        if dc > 0:
            col_mask[-dc:] = False
        elif dc < 0:
            col_mask[:-dc] = False

        valid = np.outer(row_mask, col_mask)
        drop = (src - shifted) / dist

        # Update flow_dir where this direction gives a steeper drop
        improve = valid & (drop > 0)
        if d == 0:
            best_drop = np.where(improve, drop, -np.inf)
            flow_dir = np.where(improve, d, flow_dir)
        else:
            better = improve & (drop > best_drop)
            best_drop = np.where(better, drop, best_drop)
            flow_dir = np.where(better, d, flow_dir)

    return flow_dir.astype(np.int8)


# ─── Flow accumulation ───────────────────────────────────────────────────────

def _calc_flow_accumulation(elev: np.ndarray, flow_dir: np.ndarray,
                             n_rows: int, n_cols: int) -> np.ndarray:
    """
    Flow accumulation via topological sort (highest → lowest elevation).
    Each cell starts with 1 (itself) and passes its count to its downstream neighbour.
    """
    accum = np.ones((n_rows, n_cols), dtype=np.float64)

    # Sort cells from highest to lowest
    flat_idx = np.argsort(elev.ravel())[::-1]

    for idx in flat_idx:
        row, col = divmod(int(idx), n_cols)
        d = int(flow_dir[row, col])
        if d < 0:
            continue
        nr = row + int(DR[d])
        nc = col + int(DC[d])
        if 0 <= nr < n_rows and 0 <= nc < n_cols:
            accum[nr, nc] += accum[row, col]

    return accum
